inbm/dispatcher-agent/dispatcher/source/ubuntu_source_manager.py

```diff
@@ -98,11 +101,19 @@ def __init__(self) -> None:
 
     def add(self, parameters: ApplicationAddSourceParameters) -> None:
         """Adds a source file and optional GPG key to be used during Ubuntu application updates."""
-        # Step 1: Add key (Optional)
+        # Step 1: Verify gpg key uri from trusted repo list 
         if parameters.gpg_key_name and parameters.gpg_key_uri:
+            try:
+               url = parameters.gpg_key_uri
+               #URL slicing to remove the last segment (filename) from the URL 
+               source = url.value[:-(len(url.value.split('/')[-1]) + 1)]
```

Comment:
**Suggestion:** The URL slicing logic to remove the last segment from the URL could be simplified by using urllib.parse.urlparse and manipulating the path to remove the last segment. [enhancement]
**NOTE:** Ensure that the suggested code is comprehensive. This serves as an example of the code you could use. The user should verify it after accepting the suggestion.
```suggestion
               from urllib.parse import urlparse

               parsed_url = urlparse(parameters.gpg_key_uri.value)
               path_segments = parsed_url.path.rsplit('/', 1)[0]
               source = parsed_url._replace(path=path_segments).geturl()
```